[Bug]: capture_thought hides malformed memory DB and returns generic create_failed.

### Bug type

Regression (worked before, now fails)

### Summary

  `capture_thought` can fail with the generic message `Error: failed to create memory.` even when the real fault is a
│  malformed memory store inside `.gsd/gsd.db`. Auto-mode then continues and completes the task/slice, so the
│  durable-knowledge subsystem stays broken without exposing the real SQLite error or attempting repair.

Three code paths combine to hide the real failure:
│
│  1. `src/resources/extensions/gsd/memory-store.ts:506-531` swallows every exception in `createMemory()` and returns
│  `null`.
│
│  ```ts
│  try {
│    ...
│    rewriteMemoryId(placeholder, realId);
│    return realId;
│  } catch {
│    return null;
│  }
│  ```
│
│  2. `src/resources/extensions/gsd/tools/memory-tools.ts:99-105` maps that `null` into the opaque tool result
│  `create_failed`.
│
│  ```ts
│  const id = createMemory({ category, content, confidence, scope, tags, structuredFields });
│  if (!id) {
│    return {
│      content: [{ type: "text", text: "Error: failed to create memory." }],
│      details: { operation: "memory_capture", error: "create_failed" },
│      isError: true,
│    };
│  }
│  ```
│
│  3. `src/resources/extensions/gsd/gsd-db.ts:1325-1346` only attempts malformed-DB recovery during `openDatabase()` /
│  `initSchema()`. If the DB opens successfully and later memory-table reads or writes hit `database disk image is
│  malformed`, there is no statement-time repair path.
│
│  ```ts
│  try {
│    initSchema(adapter, fileBacked);
│  } catch (err) {
│    if (fileBacked && err instanceof Error && err.message?.includes("malformed")) {
│      adapter.exec("VACUUM");
│      initSchema(adapter, fileBacked);
│      ...
│    } else {
│      throw err;
│    }
│  }
│  ```
│
│  Related behavior: `src/resources/extensions/gsd/auto-unit-closeout.ts:43-50` runs memory extraction fire-and-forget,
│  so the broken memory subsystem does not block task/slice completion and can stay degraded quietly.

### Steps to reproduce

During `complete-slice/M001/S03`, the closer reviewed task summaries and called `gsd_capture_thought` for an
│  architectural decision captured from the slice. The tool failed with `Error: failed to create memory.` and auto-mode
│  immediately continued to `gsd_complete_slice`, leaving the slice complete but the memory layer unhealthy.
│
│  The same failure pattern was already visible earlier in:
│  - `execute-task/M001/S02/T01`
│  - `execute-task/M001/S02/T02`
│
│  ## Forensic Evidence
│  - Activity log `.gsd/activity/023-complete-slice-M001-S03.jsonl` shows:
│    - a `gsd_capture_thought` call with an architecture decision payload,
│    - an error tool result `Error: failed to create memory.`,
│    - then immediate continuation into `gsd_complete_slice`.
│  - Activity logs `.gsd/activity/016-execute-task-M001-S02-T01.jsonl` and
│  `.gsd/activity/017-execute-task-M001-S02-T02.jsonl` show the same `capture_thought` failure pattern earlier in the
│  run.
│  - A checkpointed copy of `.gsd/gsd.db` reproduces the underlying SQLite fault on the memory subsystem:
│    - inserting into `memories` fails with `database disk image is malformed`
│    - `PRAGMA integrity_check` also reports `database disk image is malformed`
│  - The DB-open recovery path in `gsd-db.ts` only handles malformed errors during `initSchema()`, so later
│  statement-time corruption in memory reads/writes bypasses recovery entirely.

### Expected behavior

When the memory layer hits SQLite corruption or other write failures, GSD should:
│
│  1. **Preserve the real exception** instead of collapsing it to `null` / `create_failed`.
│     - Either let `createMemory()` throw, or return a structured `{ ok: false, error }` result.
│  2. **Surface the real cause** in `capture_thought` tool results and logs.
│     - Example: `details.error: "database disk image is malformed"`.
│  3. **Attempt one targeted repair path** for memory-table corruption detected after open time.
│     - Example: integrity-check / rebuild only the memory subsystem tables (`memories`, `memories_fts`,
│  `memory_embeddings`, `memory_relations`, `memory_processed_units`) instead of leaving the whole feature permanently
│  broken.
│  4. **Emit an explicit degraded-state warning** so forensics and operators can see that memory capture is unhealthy
│  even if task/slice completion continues.

### Actual behavior

`capture_thought` can fail with the generic message `Error: failed to create memory.` even when the real fault is a
│  malformed memory store inside `.gsd/gsd.db`. Auto-mode then continues and completes the task/slice, so the
│  durable-knowledge subsystem stays broken without exposing the real SQLite error or attempting repair.

### GSD version

2.77

### Node.js version

_No response_

### Operating system

windows 11

### Install method

npm global install

### Affected area

Slice or task execution

### Model and provider

_No response_

### Logs, screenshots, and evidence

```shell

```

### Impact

_No response_

### Additional information

_No response_

Findings from /gsd forensics
│
│  What happened
│  - The failure is real, but it is narrower than the rest of auto-mode:
│      - capture_thought started failing in later units (execute-task/M001/S02/T01, execute-task/M001/S02/T02, and again
│  in complete-slice/M001/S03).
│      - The slice still completed because memory capture is non-blocking, so GSD kept going and closed tasks/slices
│  anyway.
│  - I traced the user-visible error path:
│      - tools/memory-tools.js:62-67 turns a falsy createMemory(...) result into the generic message Error: failed to
│  create memory.
│      - memory-store.js:372-405 catches every exception inside createMemory() and returns null, discarding the real
│  SQLite error.
│  - I then inspected a checkpointed copy of .gsd/gsd.db and reproduced the underlying DB error on the memory table:
│      - insertMemoryRow(...) on the copied DB fails with database disk image is malformed.
│      - PRAGMA integrity_check on that copy also reports database disk image is malformed.
│      - Non-memory workflow state still reads, which matches your report: milestone/slice/task completion succeeded
│  while memory writes failed.
│
│  Why it happened
│  - The immediate root cause of the visible failure is an opaque error-handling chain:
│      - memory-store.js:404-405:
│        ```js
│          catch {
│              return null;
│          }
│        ```
│      - tools/memory-tools.js:62-67:
│        ```js
│          const id = createMemory({ category, content, confidence, scope, tags, structuredFields });
│          if (!id) {
│              return {
│                  content: [{ type: "text", text: "Error: failed to create memory." }],
│                  details: { operation: "memory_capture", error: "create_failed" },
│                  isError: true,
│              };
│          }
│        ```
│  - The actual underlying failure is a malformed memory store in .gsd/gsd.db, not bad tool arguments.
│  - GSD does have malformed-DB recovery, but only during DB open / schema init:
│      - gsd-db.js:1215-1237 attempts VACUUM only if initSchema(...) throws malformed.
│      - If the DB opens successfully and the corruption is hit later during a normal memory read/write, that recovery
│  path never runs.
│  - So the user gets:
│      - no SQLite error,
│      - no repair attempt,
│      - just create_failed.
│
│  Code snippet
│  - Problematic swallowing:
│      - memory-store.js:372-405
│        ```js
│          export function createMemory(fields) {
│              ...
│              try {
│                  ...
│                  rewriteMemoryId(placeholder, realId);
│                  return realId;
│              }
│              catch {
│                  return null;
│              }
│          }
│        ```
│  - Problematic masking:
│      - tools/memory-tools.js:62-67
│        ```js
│          const id = createMemory({ ... });
│          if (!id) {
│              return {
│                  content: [{ type: "text", text: "Error: failed to create memory." }],
│                  details: { operation: "memory_capture", error: "create_failed" },
│                  isError: true,
│              };
│          }
│        ```
│  - Recovery gap:
│      - gsd-db.js:1218-1225
│        ```js
│          if (fileBacked && err instanceof Error && err.message?.includes("malformed")) {
│              adapter.exec("VACUUM");
│              initSchema(adapter, fileBacked);
│              process.stderr.write("gsd-db: recovered corrupt database via VACUUM\n");
│          }
│        ```
│      - This only covers open-time/schema-time corruption, not later statement-time corruption in memories.
│
│  What it should do instead
│  - createMemory() should not swallow the real exception.
│      - Either throw it, or return { ok:false, error }.
│  - executeMemoryCapture() should surface the real SQLite message in details.error, at minimum.
│  - On malformed from memory reads/writes, GSD should attempt one repair path before giving up:
│      - run integrity check / targeted repair,
│      - or rebuild only the memory tables (memories, memories_fts, memory_embeddings, memory_relations,
│  memory_processed_units) instead of leaving the whole feature permanently broken.
│  - Also add a warning log when memory capture fails, so forensics doesn’t depend on reproducing the DB error offline.
│
│  Recovery right now
│  - You are not blocked on milestone progress:
│      - the report already shows 3/3 slices complete and 8/8 tasks complete.
│      - this bug affects the memory layer, not core task/slice completion.
│  - Short-term:
│      - treat capture_thought failure as non-fatal for this project run.
│  - If you need memory tools working immediately:
│      - back up .gsd/gsd.db first,
│      - then restore from the last known-good DB snapshot if you have one.
│  - If no backup exists, the safer fix is a code patch in GSD that can repair or rebuild the memory tables, rather than
│  manually poking the live DB.
│
│  Most relevant source references
│  - tools/memory-tools.js:62-67
│  - memory-store.js:372-405
│  - gsd-db.js:1215-1237
│  - auto-unit-closeout.js:17-25 (memory extraction is fire-and-forget, so memory-layer failures can surface later and
│  be easy to miss)

# GSD Forensic Report

**Generated:** 2026-04-25T06:35:48.530Z
**GSD Version:** 2.77.0
**Active Milestone:** M001
**Active Slice:** none
**Active Worktree:** none

## Problem Description

Tool gsd_capture_thought                                                                                        Error│  gsd_capture_thought││    "made_by": "agent",│  ...││  Error: failed to create memory.

## Anomalies Detected (5)

### [ERROR] stuck-loop: Unit execute-task/M001/S01/T01 was dispatched 8 times
- Unit: execute-task/M001/S01/T01
- Repeated dispatch within the same auto session suggests the unit completed but its artifacts were not verified, or the state machine kept returning it. Cross-session recovery runs are ignored.

### [WARNING] cost-spike: $1.46 vs $0.196 average for execute-task
- Unit: execute-task/M001/S03/T01
- Unit execute-task/M001/S03/T01 cost 7.4x the average. May indicate excessive retries or large context.

### [WARNING] cost-spike: $1.03 vs $0.196 average for execute-task
- Unit: execute-task/M001/S03/T02
- Unit execute-task/M001/S03/T02 cost 5.3x the average. May indicate excessive retries or large context.

### [WARNING] cost-spike: $0.649 vs $0.196 average for execute-task
- Unit: execute-task/M001/S03/T03
- Unit execute-task/M001/S03/T03 cost 3.3x the average. May indicate excessive retries or large context.

### [WARNING] error-trace: 2 error(s) in plan/slice-M001-S03
- Unit: plan/slice-M001-S03
- rg: .github: The system cannot find the file specified. (os error 2)
rg: package.json: The system cannot find the file specified. (os error 2)
rg: pyproject.toml: The system cannot find the file specified. (os error 2)
backend\app\core\config.py:17:    # Redis (populated by later tasks)
backend\app\
ENOENT: no such file or directory, access '.\.gsd\milestones\M001\slices\S03\S03-PLAN.md'

## Recent Units

| Type | ID | Cost | Duration | Model |
|------|-----|------|----------|-------|
| complete-slice | M001/S03 | $0.544 | 2m 58s | gpt-5.4 |
| execute-task | M001/S03/T03 | $0.649 | 3m 5s | gpt-5.4 |
| execute-task | M001/S03/T02 | $1.03 | 4m 32s | gpt-5.4 |
| execute-task | M001/S03/T01 | $1.46 | 8m 49s | gpt-5.4 |
| plan-slice | M001/S03 | $0.0000 | 8m 18s | claude-sonnet-4-6 |
| complete-slice | M001/S02 | $0.729 | 4m 3s | claude-sonnet-4.6 |
| execute-task | M001/S02/T02 | $0.0000 | 11m 55s | gpt-5.4 |
| execute-task | M001/S02/T01 | $0.0000 | 7m 17s | gpt-5.4 |
| plan-slice | M001/S02 | $0.0000 | 7m 9s | claude-sonnet-4.6 |
| complete-slice | M001/S01 | $0.0000 | 4m 52s | claude-sonnet-4.6 |

## Activity Log Traces (last 5)

### complete/slice-M001-S03 (seq 23)
- Tool calls: 18
- Files written: 0
- Errors: 0
- Last reasoning: Slice S03 complete.

All three tasks delivered: the compose runtime with health-cascade dependency ordering (T01), live pytest integration tests and Playwright browser proof without route mocks (T02),

### execute/task-M001-S03-T03 (seq 22)
- Tool calls: 21
- Files written: 0
- Errors: 0
- Last reasoning: ** — repo entry point with quick start, public URL table, layout guide, and smoke command
- **`docs/local-runtime.md`** — full operator guide: prerequisites, seeded credentials (`ka_user_plain / KaPla

### execute/task-M001-S03-T02 (seq 21)
- Tool calls: 32
- Files written: 0
- Errors: 0
- Last reasoning: test integration run and the Playwright live spec run (skippable via `--skip-playwright`).
- **`backend/tests/integration/test_runtime_surfaces.py`** and **`__init__.py`** — Already in place from T01.

### execute/task-M001-S03-T01 (seq 20)
- Tool calls: 44
- Files written: 0
- Errors: 0
- Last reasoning: at `:8080`), updated `config.py` (env-driven CORS), updated `health.py` (split liveness/readiness probes with secret-safe postgres+redis dependency checks), `backend/app/main.py` (uses settings.cors_o

### plan/slice-M001-S03 (seq 19)
- Tool calls: 60
- Files written: 3
- Errors: 2
- Last reasoning: Slice S03 planned.

## Doctor Issues

- [WARN] environment | env_env_file | .env.example exists but no .env or .env.local found — Copy .env.example to .env and fill in values | fixable: no
- [WARN] environment | env_docker | Project has Docker files but docker is not installed | fixable: no
- [WARN] environment | env_git_remote | Git remote 'origin' is unreachable — Remote: https://github.com/GBPL-egov/matrixApp-v2.0.git | fixable: no
- [WARN] project | provider_key_missing | GitHub Copilot — not configured — Run /gsd keys to authenticate | fixable: no
- [WARN] M001/S01/T01 | task_done_must_haves_not_verified | Task T01 has 5 must-haves but summary addresses only 1 | file: .gsd/milestones/M001/slices/S01/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S01/T02 | task_done_must_haves_not_verified | Task T02 has 5 must-haves but summary addresses only 4 | file: .gsd/milestones/M001/slices/S01/tasks/T02-SUMMARY.md | fixable: no
- [INFO] M001 | all_slices_done_missing_milestone_validation | All slices are done but M001-VALIDATION.md is missing — milestone is in validating-milestone phase | file: .gsd/milestones/M001/M001-VALIDATION.md | fixable: no
- [WARN] M001 | all_slices_done_missing_milestone_summary | All slices are done but M001-SUMMARY.md is missing — milestone is stuck in completing-milestone phase | file: .gsd/milestones/M001/M001-SUMMARY.md | fixable: no

## Activity Log Metadata

- Files: 23
- Total size: 4140.4 KB
- Oldest: 001-plan-milestone-M001.jsonl
- Newest: 023-complete-slice-M001-S03.jsonl

## Journal Summary

- Total entries: 95
- Distinct flows (iterations): 26
- Daily files: 2
- Date range: 2026-04-24T10:14:31.252Z — 2026-04-25T06:34:37.881Z

### Event Type Distribution

| Event Type | Count |
|------------|-------|
| iteration-start | 26 |
| dispatch-match | 23 |
| unit-start | 23 |
| unit-end | 23 |

### Recent Journal Events (last 20)

- 2026-04-25T06:05:40.335Z [dispatch-match] flow=66760b4a rule=planning → plan-slice unit=M001/S03
- 2026-04-25T06:05:40.408Z [unit-start] flow=66760b4a unit=M001/S03
- 2026-04-25T06:13:59.478Z [unit-end] flow=66760b4a unit=M001/S03
- 2026-04-25T06:15:04.548Z [iteration-start] flow=9b9bdc6c
- 2026-04-25T06:15:04.681Z [iteration-start] flow=3ae63665
- 2026-04-25T06:15:04.785Z [dispatch-match] flow=3ae63665 rule=executing → execute-task unit=M001/S03/T01
- 2026-04-25T06:15:04.851Z [unit-start] flow=3ae63665 unit=M001/S03/T01
- 2026-04-25T06:23:55.279Z [unit-end] flow=3ae63665 unit=M001/S03/T01
- 2026-04-25T06:23:56.474Z [iteration-start] flow=9ad23ac8
- 2026-04-25T06:23:56.728Z [dispatch-match] flow=9ad23ac8 rule=executing → execute-task unit=M001/S03/T02
- 2026-04-25T06:23:56.793Z [unit-start] flow=9ad23ac8 unit=M001/S03/T02
- 2026-04-25T06:28:29.938Z [unit-end] flow=9ad23ac8 unit=M001/S03/T02
- 2026-04-25T06:28:30.943Z [iteration-start] flow=22b1b30e
- 2026-04-25T06:28:31.186Z [dispatch-match] flow=22b1b30e rule=executing → execute-task unit=M001/S03/T03
- 2026-04-25T06:28:31.251Z [unit-start] flow=22b1b30e unit=M001/S03/T03
- 2026-04-25T06:31:37.705Z [unit-end] flow=22b1b30e unit=M001/S03/T03
- 2026-04-25T06:31:38.744Z [iteration-start] flow=ce4a888d
- 2026-04-25T06:31:38.987Z [dispatch-match] flow=ce4a888d rule=summarizing → complete-slice unit=M001/S03
- 2026-04-25T06:31:39.053Z [unit-start] flow=ce4a888d unit=M001/S03
- 2026-04-25T06:34:37.881Z [unit-end] flow=ce4a888d unit=M001/S03

--- a/src/resources/extensions/gsd/gsd-db.ts
+++ b/src/resources/extensions/gsd/gsd-db.ts
@@ -1429,8 +1429,8 @@ export function transaction<T>(fn: () => T): T {
     }
   }
 
-  _txDepth++;
   currentDb.exec("BEGIN");
+  _txDepth++;
   try {
     const result = fn();
     currentDb.exec("COMMIT");
@@ -1462,8 +1462,8 @@ export function readTransaction<T>(fn: () => T): T {
     }
   }
 
-  _txDepth++;
   currentDb.exec("BEGIN DEFERRED");
+  _txDepth++;
   try {
     const result = fn();
     currentDb.exec("COMMIT");

--- a/src/resources/extensions/gsd/gsd-db.ts
+++ b/src/resources/extensions/gsd/gsd-db.ts
@@ -1405,6 +1405,16 @@ export function checkpointDatabase(): void {
 
 let _txDepth = 0;
 
+/**
+ * Whether the current call is running inside an active SQLite transaction.
+ * Statement-time recovery paths (e.g. VACUUM retry on a malformed memory
+ * store) MUST gate on this — SQLite refuses VACUUM inside a transaction
+ * and would mask the original error with a secondary "cannot VACUUM" throw.
+ */
+export function isInTransaction(): boolean {
+  return _txDepth > 0;
+}
+
 export function transaction<T>(fn: () => T): T {
   if (!currentDb) throw new GSDError(GSD_STALE_STATE, "gsd-db: No database open");
 

--- a/src/resources/extensions/gsd/memory-store.ts
+++ b/src/resources/extensions/gsd/memory-store.ts
@@ -512,7 +512,7 @@ export function createMemory(fields: {
   if (!adapter) return null;
 
   try {
-    return doCreateMemory(adapter, fields);
+    return transaction(() => doCreateMemory(adapter, fields));
   } catch (err) {
     const message = err instanceof Error ? err.message : String(err);
 
@@ -522,8 +522,10 @@ export function createMemory(fields: {
     if (message.toLowerCase().includes('malformed') && !isInTransaction()) {
       try {
         adapter.prepare('VACUUM').run();
-        process.stderr.write('memory-store: recovered malformed memory store via VACUUM\n');
-        return doCreateMemory(adapter, fields);
+        const recoveryMessage = 'recovered malformed memory store via VACUUM';
+        process.stderr.write(`memory-store: ${recoveryMessage}\n`);
+        logWarning('memory-store', recoveryMessage);
+        return transaction(() => doCreateMemory(adapter, fields));
       } catch (retryErr) {
         const retryMsg = retryErr instanceof Error ? retryErr.message : String(retryErr);
         logWarning('memory-store', `VACUUM recovery for memory store failed: ${retryMsg}`);

--- a/src/resources/extensions/gsd/memory-store.ts
+++ b/src/resources/extensions/gsd/memory-store.ts
@@ -7,6 +7,7 @@ import {
   isDbAvailable,
   _getAdapter,
   transaction,
+  isInTransaction,
   insertMemoryRow,
   rewriteMemoryId,
   updateMemoryContentRow,
@@ -19,6 +20,7 @@ import {
   deleteMemoryRelationsFor,
 } from './gsd-db.js';
 import { createMemoryRelation, isValidRelation } from './memory-relations.js';
+import { logWarning } from './workflow-logger.js';
 
 // ─── Types ──────────────────────────────────────────────────────────────────
 
@@ -487,7 +489,13 @@ export function nextMemoryId(): string {
  * Insert a new memory with a race-safe auto-assigned ID.
  * Uses AUTOINCREMENT seq to derive the ID after insert, avoiding
  * the read-then-write race in concurrent scenarios (e.g. worktrees).
- * Returns the assigned ID, or null on failure.
+ * Returns the assigned ID, or null when the DB is unavailable.
+ *
+ * Throws on genuine SQL errors (corruption, missing tables, constraint
+ * violations) so callers can surface the underlying message instead of
+ * collapsing the failure to a generic "create_failed". See issue #4967 —
+ * the previous bare-catch swallowed "database disk image is malformed"
+ * errors, leaving the memory subsystem broken without any signal.
  */
 export function createMemory(fields: {
   category: string;
@@ -504,32 +512,66 @@ export function createMemory(fields: {
   if (!adapter) return null;
 
   try {
-    const now = new Date().toISOString();
-    // Insert with a temporary placeholder ID — seq is auto-assigned
-    const placeholder = `_TMP_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`;
-    insertMemoryRow({
-      id: placeholder,
-      category: fields.category,
-      content: fields.content,
-      confidence: fields.confidence ?? 0.8,
-      sourceUnitType: fields.source_unit_type ?? null,
-      sourceUnitId: fields.source_unit_id ?? null,
-      createdAt: now,
-      updatedAt: now,
-      scope: fields.scope ?? 'project',
-      tags: fields.tags ?? [],
-      structuredFields: fields.structuredFields ?? null,
-    });
-    // Derive the real ID from the assigned seq (SELECT is still fine via adapter)
-    const row = adapter.prepare('SELECT seq FROM memories WHERE id = :id').get({ ':id': placeholder });
-    if (!row) return placeholder; // fallback — should not happen
-    const seq = row['seq'] as number;
-    const realId = `MEM${String(seq).padStart(3, '0')}`;
-    rewriteMemoryId(placeholder, realId);
-    return realId;
-  } catch {
-    return null;
+    return doCreateMemory(adapter, fields);
+  } catch (err) {
+    const message = err instanceof Error ? err.message : String(err);
+
+    // Targeted recovery: a malformed memory store can sometimes be rebuilt
+    // by VACUUM. Skip when inside a transaction — SQLite refuses VACUUM
+    // there and a secondary throw would mask the real fault.
+    if (message.toLowerCase().includes('malformed') && !isInTransaction()) {
+      try {
+        adapter.prepare('VACUUM').run();
+        process.stderr.write('memory-store: recovered malformed memory store via VACUUM\n');
+        return doCreateMemory(adapter, fields);
+      } catch (retryErr) {
+        const retryMsg = retryErr instanceof Error ? retryErr.message : String(retryErr);
+        logWarning('memory-store', `VACUUM recovery for memory store failed: ${retryMsg}`);
+        // Surface the *original* malformed error — it's the actionable signal.
+        throw err;
+      }
+    }
+
+    throw err;
   }
+}
+
+function doCreateMemory(
+  adapter: NonNullable<ReturnType<typeof _getAdapter>>,
+  fields: {
+    category: string;
+    content: string;
+    confidence?: number;
+    source_unit_type?: string;
+    source_unit_id?: string;
+    scope?: string;
+    tags?: string[];
+    structuredFields?: Record<string, unknown> | null;
+  },
+): string {
+  const now = new Date().toISOString();
+  // Insert with a temporary placeholder ID — seq is auto-assigned
+  const placeholder = `_TMP_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`;
+  insertMemoryRow({
+    id: placeholder,
+    category: fields.category,
+    content: fields.content,
+    confidence: fields.confidence ?? 0.8,
+    sourceUnitType: fields.source_unit_type ?? null,
+    sourceUnitId: fields.source_unit_id ?? null,
+    createdAt: now,
+    updatedAt: now,
+    scope: fields.scope ?? 'project',
+    tags: fields.tags ?? [],
+    structuredFields: fields.structuredFields ?? null,
+  });
+  // Derive the real ID from the assigned seq (SELECT is still fine via adapter)
+  const row = adapter.prepare('SELECT seq FROM memories WHERE id = :id').get({ ':id': placeholder });
+  if (!row) return placeholder; // fallback — should not happen
+  const seq = row['seq'] as number;
+  const realId = `MEM${String(seq).padStart(3, '0')}`;
+  rewriteMemoryId(placeholder, realId);
+  return realId;
 }
 
 /**
@@ -728,8 +770,17 @@ export function applyMemoryActions(
       }
       enforceMemoryCap();
     });
-  } catch {
-    // non-fatal — transaction will have rolled back
+  } catch (err) {
+    // Non-fatal — the transaction has rolled back. We log a warning so a
+    // degraded memory subsystem (e.g. malformed store, missing tables) is
+    // visible to forensics instead of silently dropping every CREATE — see
+    // issue #4967, where this swallow combined with createMemory's bare
+    // catch hid SQLite corruption from the auto-mode flow entirely.
+    const message = err instanceof Error ? err.message : String(err);
+    logWarning(
+      'memory-store',
+      `applyMemoryActions failed (memory subsystem degraded): ${message}`,
+    );
   }
 }
 

--- a/src/resources/extensions/gsd/tests/gsd-db.test.ts
+++ b/src/resources/extensions/gsd/tests/gsd-db.test.ts
@@ -18,6 +18,8 @@ import {
   getActiveDecisions,
   getActiveRequirements,
   transaction,
+  readTransaction,
+  isInTransaction,
   _getAdapter,
   _resetProvider,
   insertMilestone,
@@ -381,6 +383,34 @@ describe('gsd-db', () => {
     assert.ok(d10 !== null, 'D010 should survive the failed transaction');
 
     closeDatabase();
+  });
+
+  test('gsd-db: failed BEGIN does not poison transaction depth', () => {
+    openDatabase(':memory:');
+    const adapter = _getAdapter()!;
+
+    const assertFailedBeginLeavesDepthClear = (label: string, fn: () => void) => {
+      adapter.exec('BEGIN');
+      try {
+        let threw = false;
+        try {
+          fn();
+        } catch {
+          threw = true;
+        }
+        assert.equal(threw, true, `${label} should surface the SQLite BEGIN failure`);
+        assert.equal(isInTransaction(), false, `${label} failed BEGIN must not leave depth active`);
+      } finally {
+        adapter.exec('ROLLBACK');
+      }
+    };
+
+    try {
+      assertFailedBeginLeavesDepthClear('transaction', () => transaction(() => undefined));
+      assertFailedBeginLeavesDepthClear('readTransaction', () => readTransaction(() => undefined));
+    } finally {
+      closeDatabase();
+    }
   });
 
   test('gsd-db: recreates missing verification evidence dedup index after removing duplicate rows', () => {

--- a/src/resources/extensions/gsd/tests/memory-store.test.ts
+++ b/src/resources/extensions/gsd/tests/memory-store.test.ts
@@ -329,3 +329,54 @@ test('memory-store: schema includes memories table', () => {
 
   closeDatabase();
 });
+
+// ═══════════════════════════════════════════════════════════════════════════
+// regression #4967 — createMemory must not silently swallow SQL errors
+// ═══════════════════════════════════════════════════════════════════════════
+
+test('memory-store: createMemory throws on memory-table SQL errors (regression #4967)', () => {
+  openDatabase(':memory:');
+
+  const adapter = _getAdapter()!;
+  // Drop FTS + dependents first to satisfy SQLite's trigger ordering, then
+  // the base memories table. IF EXISTS makes setup robust against schema
+  // versions that may not have created every dependent (e.g. embeddings).
+  adapter.prepare('DROP TABLE IF EXISTS memory_embeddings').run();
+  adapter.prepare('DROP TABLE IF EXISTS memories_fts').run();
+  adapter.prepare('DROP TABLE IF EXISTS memories').run();
+
+  // Pre-fix behaviour: returns null and the caller has no idea why.
+  // Post-fix behaviour: throws so the caller can surface the real SQL message.
+  assert.throws(
+    () => createMemory({ category: 'gotcha', content: 'broken store' }),
+    /memories|no such table/i,
+    'createMemory must surface SQL errors instead of returning null',
+  );
+
+  closeDatabase();
+});
+
+test('memory-store: applyMemoryActions stays non-fatal when memory store is broken (regression #4967)', () => {
+  openDatabase(':memory:');
+
+  const adapter = _getAdapter()!;
+  // Drop FTS + dependents first to satisfy SQLite's trigger ordering, then
+  // the base memories table. IF EXISTS makes setup robust against schema
+  // versions that may not have created every dependent (e.g. embeddings).
+  adapter.prepare('DROP TABLE IF EXISTS memory_embeddings').run();
+  adapter.prepare('DROP TABLE IF EXISTS memories_fts').run();
+  adapter.prepare('DROP TABLE IF EXISTS memories').run();
+
+  // applyMemoryActions wraps createMemory in a transaction with an outer
+  // catch. Even with createMemory now throwing, applyMemoryActions must not
+  // crash the auto-mode flow that calls it (memory extraction is best-effort).
+  const actions: MemoryAction[] = [
+    { action: 'CREATE', category: 'gotcha', content: 'inside-transaction call' },
+  ];
+  assert.doesNotThrow(
+    () => applyMemoryActions(actions),
+    'applyMemoryActions must absorb thrown errors so callers continue',
+  );
+
+  closeDatabase();
+});

--- a/src/resources/extensions/gsd/tests/memory-store.test.ts
+++ b/src/resources/extensions/gsd/tests/memory-store.test.ts
@@ -5,6 +5,11 @@ import {
   _getAdapter,
 } from '../gsd-db.ts';
 import {
+  _resetLogs,
+  peekLogs,
+  setStderrLoggingEnabled,
+} from '../workflow-logger.ts';
+import {
   getActiveMemories,
   getActiveMemoriesRanked,
   nextMemoryId,
@@ -354,6 +359,78 @@ test('memory-store: createMemory throws on memory-table SQL errors (regression #
   );
 
   closeDatabase();
+});
+
+test('memory-store: VACUUM retry rolls back partial memory and logs recovery', () => {
+  openDatabase(':memory:');
+
+  const adapter = _getAdapter()!;
+  const originalPrepareMethod = adapter.prepare;
+  const originalPrepare = adapter.prepare.bind(adapter);
+  const previousStderrLogging = setStderrLoggingEnabled(false);
+  const streamAny = process.stderr as unknown as {
+    write: (chunk: string | Uint8Array, ...rest: unknown[]) => boolean;
+  };
+  const originalStderrWrite = streamAny.write.bind(streamAny);
+  let selectFailures = 0;
+  let vacuumRuns = 0;
+  _resetLogs();
+
+  adapter.prepare = ((sql: string) => {
+    if (sql === 'SELECT seq FROM memories WHERE id = :id' && selectFailures === 0) {
+      const stmt = originalPrepare(sql);
+      return {
+        run: (...params: unknown[]) => stmt.run(...params),
+        get: (..._params: unknown[]) => {
+          selectFailures++;
+          throw new Error('database disk image is malformed');
+        },
+        all: (...params: unknown[]) => stmt.all(...params),
+      };
+    }
+
+    if (sql === 'VACUUM') {
+      const stmt = originalPrepare(sql);
+      return {
+        run: (...params: unknown[]) => {
+          vacuumRuns++;
+          return stmt.run(...params);
+        },
+        get: (...params: unknown[]) => stmt.get(...params),
+        all: (...params: unknown[]) => stmt.all(...params),
+      };
+    }
+
+    return originalPrepare(sql);
+  }) as typeof adapter.prepare;
+  streamAny.write = (): boolean => true;
+
+  try {
+    const id = createMemory({ category: 'gotcha', content: 'recover without duplicate' });
+    assert.equal(id, 'MEM001', 'retry should create a single first memory');
+
+    const rows = adapter.prepare('SELECT id FROM memories ORDER BY seq').all();
+    assert.deepStrictEqual(
+      rows.map((row) => row['id']),
+      ['MEM001'],
+      'failed first attempt should not leave a live _TMP_ memory behind',
+    );
+    assert.equal(selectFailures, 1, 'test should simulate one malformed SELECT after INSERT');
+    assert.equal(vacuumRuns, 1, 'malformed recovery should run VACUUM once');
+    assert.ok(
+      peekLogs().some((entry) =>
+        entry.component === 'memory-store' &&
+        entry.message === 'recovered malformed memory store via VACUUM'
+      ),
+      'successful VACUUM recovery should be emitted to the workflow logger',
+    );
+  } finally {
+    adapter.prepare = originalPrepareMethod;
+    streamAny.write = originalStderrWrite;
+    setStderrLoggingEnabled(previousStderrLogging);
+    _resetLogs();
+    closeDatabase();
+  }
 });
 
 test('memory-store: applyMemoryActions stays non-fatal when memory store is broken (regression #4967)', () => {

--- a/src/resources/extensions/gsd/tests/memory-tools.test.ts
+++ b/src/resources/extensions/gsd/tests/memory-tools.test.ts
@@ -1,7 +1,7 @@
 import { test } from 'node:test';
 import assert from 'node:assert/strict';
 
-import { closeDatabase, openDatabase } from '../gsd-db.ts';
+import { _getAdapter, closeDatabase, openDatabase } from '../gsd-db.ts';
 import { createMemory, supersedeMemory } from '../memory-store.ts';
 import {
   executeGsdGraph,
@@ -292,4 +292,36 @@ test('memory-tools: gsd_graph errors when DB is closed', () => {
   const result = executeGsdGraph({ mode: 'query', memoryId: 'MEM001' });
   assert.ok(result.isError);
   assert.equal(result.details.error, 'db_unavailable');
+});
+
+// ═══════════════════════════════════════════════════════════════════════════
+// regression #4967 — capture_thought must surface real SQL errors
+// ═══════════════════════════════════════════════════════════════════════════
+
+test('memory-tools: capture_thought surfaces underlying SQL error (regression #4967)', () => {
+  openDatabase(':memory:');
+
+  // Simulate the real-world failure mode where the memories table is gone.
+  // The original bug was a "database disk image is malformed" on the memory
+  // store; dropping the table produces a deterministic statement-time SQL
+  // error of the same shape — a thrown sqlite error during INSERT.
+  const adapter = _getAdapter()!;
+  adapter.prepare('DROP TABLE IF EXISTS memory_embeddings').run();
+  adapter.prepare('DROP TABLE IF EXISTS memories_fts').run();
+  adapter.prepare('DROP TABLE IF EXISTS memories').run();
+
+  const result = executeMemoryCapture({
+    category: 'gotcha',
+    content: 'should reveal the real reason',
+  });
+
+  assert.ok(result.isError, 'broken store should produce an error result');
+  assert.equal(result.details.operation, 'memory_capture');
+
+  const err = String(result.details.error ?? '');
+  assert.notEqual(err, 'create_failed', 'must not collapse to opaque create_failed');
+  assert.ok(err.length > 0, 'error detail must be populated');
+  assert.match(err, /memories|no such table/i, 'error must reference the underlying SQL fault');
+
+  closeDatabase();
 });

--- a/src/resources/extensions/gsd/tools/memory-tools.ts
+++ b/src/resources/extensions/gsd/tools/memory-tools.ts
@@ -96,8 +96,24 @@ export function executeMemoryCapture(params: MemoryCaptureParams): ToolExecution
   const tags = normalizeTags(params.tags);
 
   const structuredFields = normalizeStructuredFields(params.structuredFields);
-  const id = createMemory({ category, content, confidence, scope, tags, structuredFields });
+  let id: string | null;
+  try {
+    id = createMemory({ category, content, confidence, scope, tags, structuredFields });
+  } catch (err) {
+    // Surface the underlying SQL message (e.g. "database disk image is
+    // malformed", "no such table: memories") so the operator gets the
+    // actionable signal instead of an opaque "create_failed". See #4967.
+    const message = err instanceof Error ? err.message : String(err);
+    return {
+      content: [{ type: "text", text: `Error: failed to create memory: ${message}` }],
+      details: { operation: "memory_capture", error: message },
+      isError: true,
+    };
+  }
   if (!id) {
+    // DB unavailable or adapter missing — distinct from the SQL-error path
+    // above. Keep the legacy create_failed token here so any consumers that
+    // explicitly key on the unavailable case continue to work.
     return {
       content: [{ type: "text", text: "Error: failed to create memory." }],
       details: { operation: "memory_capture", error: "create_failed" },

--- a/src/resources/extensions/gsd/workflow-logger.ts
+++ b/src/resources/extensions/gsd/workflow-logger.ts
@@ -66,6 +66,7 @@ export type LogComponent =
   | "memory-embeddings" // Memory layer embedding generation
   | "memory-ingest"     // Memory layer ingestion pipeline
   | "memory-backfill"   // ADR-013: decisions->memories backfill
+  | "memory-store"      // Memory CRUD layer — surfaces SQLite/store-level faults (#4967)
   | "context-mode"     // Context-mode exec sandbox and compaction snapshot
   | "preflight"        // Clean-root preflight gate at milestone completion
   | "postUnit";     // Post-unit processing (abandon detection, overrides)

